### PR TITLE
Update file type detection

### LINKED ISSUE #54: 'rclone copy' doesn't download all of my files
I just did `rclone copy permanent-prod:/ ./permanent-down`, but I didn't get all of the files from my (one) archive.  I'll show a full transcript of the session later, but first here's the TL;DR:

I *should* have gotten the following file tree (folders are indicated with a trailing "/" -- the "/" is not actually in the name, of course):

```
archives/
  kfogel OTS test (13539)/
    My Files/
      dennys-fake-maple-syrup-2022-05-11.jpg  
      Screenshot from 2022-07-12 14-17-51.png   
      tex-kerning.png   
      worker-digging.png
      Cables I have known/
        50-foot-VGA-cable.jpg
      GNOME Workspace Switcher bug/  
        gnome-3-shell-42.0-workspace-switcher-bottom-indicator-ws1-to-ws2-2022-04-10.png  
        gnome-3-shell-42.0-workspace-switcher-bottom-indicator-ws2-to-ws1-2022-04-10.png
        This subfolder is just to, you know, test subfolders./
          uncle-neatest-whiskey-weird-endorsement.jpg
        This is an empty subfolder./
```

But instead I got this file tree:

```
archives/
  kfogel OTS test (13539)/
    My Files/
      dennys-fake-maple-syrup-2022-05-11.jpg  
      Screenshot from 2022-07-12 14-17-51.png   
      tex-kerning.png   
      worker-digging.png
      Cables I have known/
        50-foot-VGA-cable.jpg
      GNOME Workspace Switcher bug/  
        gnome-3-shell-42.0-workspace-switcher-bottom-indicator-ws1-to-ws2-2022-04-10.png  
        gnome-3-shell-42.0-workspace-switcher-bottom-indicator-ws2-to-ws1-2022-04-10.png
```

Notice how the two subfolders of the `GNOME Workspace Switcher bug/` folder are missing, and thus the JPG image file in one of those subfolders is also missing.

(By the way, not that it matters for this bug report, but yes, the name of that missing JPG image has a typo.  In my actual archive I've changed the "name" of that record, but of course that doesn't affect the downloadName, which still has the original name.  Later I may just delete the file and re-upload it with a fixed name, but I didn't want to introduce that variable while reporting this issue.)

So:

Are the subfolders missing maybe because their names end with dot?

Here's a transcript of my complete session -- notice the interesting-and-possibly-relevant error messages:

```
$ cat ~/.config/rclone/rclone.conf 
[permanent-prod]
type = sftp
host = sftp.permanent.org
user = kfogel@opentechstrategies.com
ask_password = true

$ time rclone copy permanent-prod:/ ./permanent-down
Enter SFTP password: *************************************

<3>ERROR : archives/kfogel OTS test (13539)/My Files/GNOME Workspace Switcher bug/This subfolder is just to, you know, test subfolders.: error reading source directory: List failed: is a file not a directory
<3>ERROR : archives/kfogel OTS test (13539)/My Files/GNOME Workspace Switcher bug/This is an empty subfolder.: error reading source directory: List failed: is a file not a directory
<3>ERROR : Attempt 1/3 failed with 3 errors and: march failed with 2 error(s): first error: List failed: is a file not a directory
<3>ERROR : archives/kfogel OTS test (13539)/My Files/GNOME Workspace Switcher bug/This subfolder is just to, you know, test subfolders.: error reading source directory: List failed: is a file not a directory
<3>ERROR : archives/kfogel OTS test (13539)/My Files/GNOME Workspace Switcher bug/This is an empty subfolder.: error reading source directory: List failed: is a file not a directory
<3>ERROR : Attempt 2/3 failed with 3 errors and: march failed with 2 error(s): first error: List failed: is a file not a directory
<3>ERROR : archives/kfogel OTS test (13539)/My Files/GNOME Workspace Switcher bug/This subfolder is just to, you know, test subfolders.: error reading source directory: List failed: is a file not a directory
<3>ERROR : archives/kfogel OTS test (13539)/My Files/GNOME Workspace Switcher bug/This is an empty subfolder.: error reading source directory: List failed: is a file not a directory
<3>ERROR : Attempt 3/3 failed with 3 errors and: march failed with 2 error(s): first error: List failed: is a file not a directory

### It just hung here for many minutes; eventually I did ^C to interrupt and exit.
### That's probably known bug https://github.com/PermanentOrg/sftp-service/issues/45.

real	11m2.034s
user	0m0.533s
sys	0m0.539s
$ 
```

If anyone wants the password to my account in order to reproduce this, just let me know; there's nothing private in there.

For your amusement, I will attach the missing JPG image file here (using the new user-visible name that it now shows in the Permanent web UI).  I think you'll agree that there is indeed a very odd claim being made on the front of that bottle.

weird-whiskey-endorsement

### PR DESCRIPTION
**Pull request by slifty**
This PR implements new logic for path type detection (e.g. file vs folder).  This fixes issues where folders with `.` in the name would be incorrectly treated as a file.

Resolves #54

### CODE REVIEW

File: src/classes/PermanentFileSystem.ts [hunk 1]
```diff
@@ -161,6 +192,12 @@ export class PermanentFileSystem {
       return cachedFolder;
     }
 
+    if (isRootPath(requestedPath)
```
Comment (slifty) [inline src/classes/PermanentFileSystem.ts:195]: It's probably better to just check `isItemPath` than check all three of these explicitly.

File: src/classes/PermanentFileSystem.ts [hunk 2]
```diff
@@ -34,6 +34,11 @@ const isArchivePath = (requestedPath: string): boolean => (
   && requestedPath.split('/').length === 3
 );
 
+const isArchiveRootLevelFolderPath = (requestedPath: string): boolean => (
+  requestedPath.startsWith('/archives')
+  && requestedPath.split('/').length === 4
```
Comment (Fenn-CS) [inline src/classes/PermanentFileSystem.ts:39]:
```php
requestedPath.startsWith('/archives') && requestedPath.split('/').length === 4
```

For the sake of this review and perhaps for other references can you please explain the splitted path lengths, 4, 3 ... as seen here and below?

File: src/classes/PermanentFileSystem.ts [hunk 3]
```diff
@@ -72,6 +77,32 @@ export class PermanentFileSystem {
...
+    );
+    if (targetIsRecord) {
+      return fs.constants.S_IFREG;
+    }
+    const targetIsFolder = parentFolder.folders.some(
+      (folder) => folder.name === childName,
+    );
+    if (targetIsFolder) {
+      return fs.constants.S_IFDIR;
+    }
```
Comment (Fenn-CS) [inline src/classes/PermanentFileSystem.ts:102]: This make sense!

**Review by Fenn-CS — APPROVED** [review]
This looks okay to me. However, I am not sure why we have to do all these sftp-service side, I mean? Could we find the way for the SDK to just tell us the answer when we ask

**Comment by slifty** [comment]
The SDK doesn't have any way of mapping an SFTP file path to a permanent record; SFTP is responsible for simulating the filesystem and mapping SDK calls to that simulated filesystem.  We could some day explore the idea of releasing the PermanentFilesystem as its own package (a virtual file system) but if / when we do that it should be treated as its own project and should be done in a way that mirrors expectations of filesystems (as opposed to just being a utility class within this service)

We *are* using the SDK to answer the question though -- loading folders and records are SDK calls.

File: src/classes/PermanentFileSystem.ts (hunk 2 above)
Comment (Fenn-CS) [reply src/classes/PermanentFileSystem.ts:39]: +1, I would recommend the comments.

Comment (slifty) [reply src/classes/PermanentFileSystem.ts:41]: I've added those comments!

File: src/classes/PermanentFileSystem.ts (hunk 3 above)
Comment (slifty) [reply src/classes/PermanentFileSystem.ts:105]: I just saw the second part of this!  I replied below in the main thread (copying the important concept here for posterity: basically the SDK doesn't understand the concept of a file path / doesn't think in terms of file system.)

**Pull request merged**